### ISSUE TITLE
[Android] Aging devices: minSdkVersion API level recommendation

### ISSUE THREAD

**Issue by tdeffains** [green-code-initiative/creedengo-challenge#40]
# Aging devices: minSdkVersion API level recommendation
## Platform
| OS | OS version | Langage |
|----------|------------|-----------|
| Android | - | Gradle |
## Main caracteristics
| ID | Title | Category | Sub-category |
|---------|----------------------------------|-------------|----------------|
| - | minSdkVersion API level recommendation | Environment | Aging devices |
## Severity / Remediation Cost
| Severity | Remediation Cost |
|----------|------------------|
| - | - |
## Rule short description
Set application minSdkVersion to high API level will make users' devices obsolete regarding the application.
### Text
As per Android Studio API Version Distribution minimum API Level 21 reaches 99,3% of devices
Also Jetpack Compose is now the reference for building UI and requires API Level 21

![Android Studio API Version Distribution](https://user-images.githubusercontent.com/28409630/230114590-9d892df7-6cd3-4580-99fc-651ed7bbf2d3.png)
https://apilevels.com/
https://developer.android.com/jetpack/compose/setup
### build.gradle
```gradle
defaultConfig {
        minSdkVersion 21
}
```
## Implementation principle
- Retrieve application **_build.gradle_** file
- Extract **_android {}_** bloc
- Extract **_defaultConfig {}_** bloc
- Check if the **_minSdkVersion_** or **_minSdk_** version set (warning: it may be set to some hardcoded value set in the root **_build.gradle_** file) is higher than 21
https://developer.android.com/ndk/guides/sdk-versions#minsdkversion
### Variant (open discussion)
Previous implementation may change as a new Android version is released each year (with incremented API level)
- Repeat previous steps from 1 to 3
- Instead of comparing to 21, compare to the latest API level minus **[static number - to be defined/discussed]**
   - Latest API level can be retrieved [here](https://developer.android.com/tools/releases/platforms) (do not take into account _Developer Preview_ versions)
   - **Warning - how to determine [static number] knowing that Google may release several versions per year ([history](https://apilevels.com/))**

**Comment by olegoaer** [comment]
I agree. Maybe **scraping** from the https://apilevels.com/ website enable to determine the ideal `minSdkVersion`

I don't know if embedding a scraping step in a sonar detection rule is commonplace. But let's try!

**Comment by olegoaer** [comment]
Warning: Remote calls are harmful for the security level of the plugin

**Comment by tdeffains** [comment]
Updating API levels as of today (20/05/25)

<img width="1096" alt="Image" src="https://github.com/user-attachments/assets/e9ad1650-b3e6-4ae5-bbc6-fb8ceecba379" />

Proposal: define 3 severity levels depending on 3 minSdkVersion threshold
- API level 21 (Android 5.0)
   - Supports 99.9% of devices (drops 0.1%)
   - Material design support
   - [Jetpack](https://developer.android.com/jetpack)/[AndroidX](https://developer.android.com/jetpack/androidx) libraries [require](https://developer.android.com/jetpack/androidx/versions#version-table) a minSdk of 21 or higher since April 2024.
   - [Jetpack Compose](https://developer.android.com/jetpack/compose) requires a minSdk of 21 or higher.
   - Google Play services v23.30.99+ (August 2023) [drops support](https://android-developers.googleblog.com/2023/07/google-play-services-discontinuing-updates-for-kitkat.html) for API levels below 21.
- API level 24 (Android 7.0)
   - Supports 98.6% of devices (drops 1.4%)
   - Several security features support (Key Attestation, Network Security Config, Default Trusted Certificate Authority)
   - Battery life features support
- API level 29 (Android 10)
   - Supports 87.6% of devices (drops 12.4%)
   - Several security and privacy features (Storage Encryption, TLS 1.3 by Default, etc)
   - Dark theme support

29 < minSdkLevel -> Severity Critical
24 < minSdkLevel <= 29 -> Severity Major
21 < minSdkLevel <= 24 -> Severity Minor
(minSdkLevel <= 21 -> No issue)

**Comment by tdeffains** [comment]
EDIT after discussion: it is not possible on Sonar to implement different severity levels for same rule
I'll post below new rule description with check on minSdkLevel 29 support (Android 10)

**Comment by tdeffains** [comment]
# Aging devices: minSdkVersion API level 29 recommendation
## Platform
| OS | OS version | Langage |
|----------|------------|-----------|
| Android | - | Gradle |
## Main caracteristics
| ID | Title | Category | Sub-category |
|---------|----------------------------------|-------------|----------------|
| - | minSdkVersion API level recommendation | Environment | Longevity |
## Severity / Remediation Cost
| Severity | Remediation Cost |
|----------|------------------|
| Critical | 5d |
## Rule short description
Set application minSdkVersion to high API level will make users' devices obsolete regarding the application.
## Rule complete description
### Text
As per Android Studio API Version Distribution minimum API Level 29 (Android 10) reaches 87,6% of devices

<img width="1095" alt="Image" src="https://github.com/user-attachments/assets/8dc7311b-40eb-403d-8b44-092989e4096a" />

https://apilevels.com/

### build.gradle
```gradle
defaultConfig {
        minSdkVersion 29
}
```

## Implementation principle
- Retrieve application **_build.gradle_** file
- Extract **_android {}_** bloc
- Extract **_defaultConfig {}_** bloc
- Check if the **_minSdkVersion_** or **_minSdk_** version set (warning: it may be set to some hardcoded value set in the root **_build.gradle_** file) is higher than 29
https://developer.android.com/ndk/guides/sdk-versions#minsdkversion

## Example
